### PR TITLE
brotli: ensure CMAKE_SYSTEM_NAME=Windows on mingw cross builds

### PR DESCRIPTION
**Pull request by mcmtroffaes**
###### Motivation for this change

CMAKE passes the flag -rdynamic otherwise (which is not recognized by mingw). Setting the appropriate flag avoids this problem and allows the build to succeed.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions (Fedora)
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

I've also tested the resulting executable using wine on fedora (compressing & decompressing a file, checked the difference).

@vcunat @Ericson2314

### CODE REVIEW

**Comment by mcmtroffaes** [comment]
For reference, here's the shell.nix file I used for testing:

```nix
with import <nixpkgs> {};
{
  my-env = pkgs.stdenv.mkDerivation {
    name = "my-env";
    buildInputs = [
      pkgs.pkgsCross.mingwW64.brotli
    ];
  };
}
```

**Comment by matthewbauer** [comment]
Thanks! I'm glad someone is also working with Mingw.

I think we can actually set this up for all CMake builds. Reading the documentation, it looks like by default CMAKE_SYSTEM_NAME is set to CMAKE_HOST_SYSTEM_NAME (which awkwardly is a different kind of "host" compared to the one in Nixpkgs): https://cmake.org/cmake/help/latest/variable/CMAKE_SYSTEM_NAME.html

The only thing is keeping track of ```uname -s``` output for each system. @Ericson2314 is the something we can do? (otherwise we can probably just map "kernel" to system name).

https://github.com/NixOS/nixpkgs/blob/92a047a6c4d46a222e9c323ea85882d0a7a13af8/pkgs/development/tools/build-managers/cmake/setup-hook.sh#L32-L33

**Comment by matthewbauer** [comment]
@GrahamcOfBorg build pkgs.pkgsCross.mingwW64.brotli

**Comment by Ericson2314** [comment]
Yeah `CMAKE_SYSTEM_*` looks like Autoconf's and our host platform, and `CMAKE_HOST_SYSTEM_*` looks like Autoconf's and our build platform. If I understand https://gitlab.kitware.com/cmake/community/wikis/doc/cmake/CrossCompiling correctly, those two might have the same structure, though they only spell it out explicitly for `*_NAME`.

**Review by matthewbauer — APPROVED** [review]
Looks okay to merge right now (not sure if it should be master or staging though). Discussing with @ericson2314 how we can do this everywhere - it gets a little complicated because now we need some kind of a wrapper to hold what the targeted system is, previously we reused the same thing every time. Anyway, it's used in other places already so should be okay.

**Comment by Ericson2314** [comment]
Maybe I would add this version to 18.09, and do the wrapper on maste/staging.

**Comment by matthewbauer** [comment]
I am still wanting to explore ways to avoid the wrapping. The setup hook should have enough information to get the target from the c compiler without a wrapper. Somehow parsing dumpmachine output? Or some other variable

**Comment by Ericson2314** [comment]
I just don't see how computing the same information every time dynamically is superior to doing it once statically. The wrapper would be a lot simpler than cc-wrapper.

**Comment by Ericson2314** [comment]
I agree. Merge this now and generalize after.

**Pull request merged**
